### PR TITLE
refactor(api): Change Authority to Authorities[] in checkAuthorityOverWorkspace

### LINKED ISSUE #196: API: Update `authority-checker.service.ts` to allow checking for multiple authorities
## Description

The authority check input in `authority-checker.service.ts` file currently takes the following interface as its input type:
```ts
AuthorityInput {
  userId: string
  entity: {
    id?: string
    name?: string
  }
  authority: Authority
  prisma: PrismaClient
}
```

The problem with this approach is, we are not able to check multiple authorities at once. Ideally, we would want `authority` to be renamed to `...authorities: Authority[]`. This will allow us to pass multiple authorities so that we can check them at once. 

Ideally, we would like the comparison in the functions to be set difference operations.

## Solution

- Refactor the `authority` field and all its occurrences
- Then, pass the required authorities and collective authorities into this function to check for any overlap.
- Add relevant tests for each of project, environment, workspace, secret, variable and approval in individual `.e2e.spec.ts` files in the modules

### PR DESCRIPTION
**Pull request by anaskhan28**
## Description
Changed the authority check from a single `Authority` to an array of `Authorities[]` in the `checkAuthorityOverWorkspace` method. This allows for more flexible permission checks where a user can have multiple authorities.

Fixes #196

## Dependencies
No new dependencies were added.

## Future Improvements
- Consider implementing a more granular authority system if needed in the future.
- Potentially add a caching mechanism for frequently checked authorities to improve performance.

## Mentions
@[relevant-team-member] @[project-manager]

## Screenshots of relevant screens
N/A - This change is backend-only and does not affect the UI.

## Developer's checklist
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**
- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [x] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [ ] There are no UI/UX issues

### Documentation Update
- [x] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

Note: Please review and update the documentation to reflect the change from `Authority` to `Authorities[]`.

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Refactored multiple services to change the authority check from a single `Authority` to an array of `Authorities[]`.
- Updated logic in `AuthorityCheckerService` to handle multiple authorities.
- Improved error messages for unauthorized access across various services.
- Ensured all relevant service methods now accept and process an array of authorities.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>authority-checker.service.ts</strong><dd><code>Refactor authority checks to support multiple authorities</code></dd></summary>
<hr>

apps/api/src/common/authority-checker.service.ts

<li>Changed <code>authority</code> parameter to <code>authorities</code> array in multiple methods.<br> <li> Updated logic to check for multiple authorities.<br> <li> Improved error messages for unauthorized access.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-9dc32989afcb1043bc772670df8a1c993dae9721b8b44a8df13022471cce987e">+77/-79</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>environment.service.ts</strong><dd><code>Update environment service to support multiple authorities</code></dd></summary>
<hr>

apps/api/src/environment/service/environment.service.ts

<li>Updated calls to <code>checkAuthorityOverProject</code> and <br><code>checkAuthorityOverEnvironment</code> to use <code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-658d3ecf90efc583248d280885f43ee30a341ca9949c2a48c1312d3ed6f3acbd">+5/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>event.service.ts</strong><dd><code>Update event service to support multiple authorities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/event/service/event.service.ts

<li>Updated call to <code>checkAuthorityOverWorkspace</code> to use <code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-bdcc00beac2cc963775cf8d6fc5a1c7ad5a1a179b374063592889b20c1aeb8fc">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>integration.service.ts</strong><dd><code>Update integration service to support multiple authorities</code></dd></summary>
<hr>

apps/api/src/integration/service/integration.service.ts

<li>Updated calls to <code>checkAuthorityOverWorkspace</code>, <br><code>checkAuthorityOverProject</code>, and <code>checkAuthorityOverEnvironment</code> to use <br><code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-dc7461d8d122f2cd2689e572229ca373cf9d2e7afd0104f67b513e11aa019a4e">+10/-10</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>project.service.ts</strong><dd><code>Update project service to support multiple authorities</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/project/service/project.service.ts

<li>Updated calls to <code>checkAuthorityOverWorkspace</code> and <br><code>checkAuthorityOverProject</code> to use <code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-d85f483eff15c28a4b3e5f7a54c9b159c71003f8a95ef7a4be00eb80ec72b2fa">+12/-12</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>secret.service.ts</strong><dd><code>Update secret service to support multiple authorities</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/secret/service/secret.service.ts

<li>Updated calls to <code>checkAuthorityOverProject</code>, <br><code>checkAuthorityOverEnvironment</code>, and <code>checkAuthorityOverSecret</code> to use <br><code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-dee38177617754972e3d3f727d7f1536566d7a784b7ffdda74aa97d6eec4cbc5">+9/-9</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>change-notifier.socket.ts</strong><dd><code>Update change notifier to support multiple authorities</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/socket/change-notifier.socket.ts

<li>Updated calls to <code>checkAuthorityOverWorkspace</code>, <br><code>checkAuthorityOverProject</code>, and <code>checkAuthorityOverEnvironment</code> to use <br><code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-5afa9c3cdf1140526f7b0b1d5163c7be1ff61241e67b6ff008095ca40da31da9">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variable.service.ts</strong><dd><code>Update variable service to support multiple authorities</code>&nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/variable/service/variable.service.ts

<li>Updated calls to <code>checkAuthorityOverProject</code>, <br><code>checkAuthorityOverEnvironment</code>, and <code>checkAuthorityOverVariable</code> to use <br><code>authorities</code> array.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-db1895137a9c036373abefb8c924150d4f2022fd6d4f5271cc6c95c443bf01cf">+9/-9</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>workspace-role.service.ts</strong><dd><code>Update workspace role service to support multiple authorities</code></dd></summary>
<hr>

apps/api/src/workspace-role/service/workspace-role.service.ts

<li>Updated calls to <code>checkAuthorityOverWorkspace</code> to use <code>authorities</code> array.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-fa0a0e5d18d758148d81c6236ae88c9e4e117214826e1a03cc9cc0441d111320">+5/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>workspace.service.ts</strong><dd><code>Update workspace service to support multiple authorities</code>&nbsp; </dd></summary>
<hr>

apps/api/src/workspace/service/workspace.service.ts

<li>Updated calls to <code>checkAuthorityOverWorkspace</code> to use <code>authorities</code> array.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/401/files#diff-a92613f23cf2834d52eaedb87c4e8c460d48b6fca8df4ccc968b2a8af19a9c3f">+12/-12</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: apps/api/src/common/authority-checker.service.ts [hunk 1]
```diff
@@ -68,13 +67,15 @@ export class AuthorityCheckerService {
...
     )
 
-    // Check if the user has the authority to perform the action
-    if (
-      !permittedAuthorities.has(authority) &&
-      !permittedAuthorities.has(Authority.WORKSPACE_ADMIN)
-    ) {
+    const hasRequiredAuthority = authorities.some(
+      (auth) =>
+        permittedAuthorities.has(auth) ||
```
Comment (rajdip-b) [inline apps/api/src/common/authority-checker.service.ts:72]: As per the new logic, we would need all the authorities passed into `authorities` to be contained in `permittedAuthorites`. So, some operation wont work. 

You can add these two steps to improve the performance:
- Check for workspace admin role before the loop
- Use set operations for authority check (IDK how that can be implemented, but feel free to skip it if its too challenging.)

**Review by rajdip-b — COMMENTED** [review]
I've left a detailed explaination about the change in one function. I would need you to extract the bit that checks the actual authorities into a private function and then calling that function from the other functions.

File: apps/api/src/common/authority-checker.service.ts (hunk 1 above)
Comment (anaskhan28) [reply apps/api/src/common/authority-checker.service.ts:72]: sure I'll try to do it 🫡

Comment (rajdip-b) [reply apps/api/src/common/authority-checker.service.ts:72]: @anaskhan28 any progress on this?

Comment (anaskhan28) [reply apps/api/src/common/authority-checker.service.ts:72]: I was busy with college will probably try to complete it by tomorrow 🫡

Comment (rajdip-b) [reply apps/api/src/common/authority-checker.service.ts:72]: @anaskhan28 hey! would you please be able to make the fixes? I would really love to merge the changes!

**Comment by rajdip-b** [comment]
Hey bro, I was fixing your PR, but then something went wrong i guess and then git didnt pick up your changes and marked the PR invalid. Although, the contribution was valid and I have merged the code in our `develop` branch. You can refer to this screenshot:
image

**Pull request closed without merging**